server/conn.go

```diff
@@ -953,9 +963,17 @@ func (cc *clientConn) checkAuthPlugin(ctx context.Context, resp *handshakeRespon
 	return nil, nil
 }
 
-func (cc *clientConn) PeerHost(hasPassword string) (host, port string, err error) {
+func (cc *clientConn) PeerHost(hasPassword string, update bool) (host, port string, err error) {
+	// already get peer host
 	if len(cc.peerHost) > 0 {
-		return cc.peerHost, cc.peerPort, nil
+		// Proxy protocol enabled and not update
+		if cc.ppEnabled && !update {
+			return cc.peerHost, cc.peerPort, nil
+		}
+		// Proxy protocol not enabled
+		if !cc.ppEnabled {
+			return cc.peerHost, cc.peerPort, nil
+		}
```

Comment:
The peerHost and serverHost may be parsed more than once even for the same address. If you think the overhead is ignorable, you can just remove this judgement.

If you think it's not, you can add another member, such as `lastPeerAddr string` to ensure it's parsed only once, and you can remove the parameter `update bool`.

Just a suggestion, it's up to you.

Reply:
I change the logic, just update after first packet read in handshake function call.